File: src/hbse/core/records.py
"""Encrypted secret record models."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from hbse.core.models import SecretStatus, SecretType
from hbse.core.serialization import b64url_decode_no_padding, b64url_no_padding


class SecretRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["1"] = "1"
    vault_id: str
    namespace_id: str
    secret_id: str
    secret_ref: str
    secret_version: int = Field(ge=1)
    status: SecretStatus = SecretStatus.ACTIVE
    secret_type: SecretType = SecretType.GENERIC
    algorithm_id: Literal["AES-256-GCM"] = "AES-256-GCM"
    wrap_algorithm_id: Literal["AES-256-GCM"] = "AES-256-GCM"
    policy_id: str = "default-deny"
    policy_hash: str
    metadata_hash: str
    provider_policy_hash: str = "unbound-provider-policy"
    created_at: datetime
    secret_nonce: str
    ciphertext: str
    dek_nonce: str
    wrapped_dek: str
    secret_aad_hash: str
    dek_wrap_aad_hash: str

    @classmethod
    def encode_bytes(cls, value: bytes) -> str:
        return b64url_no_padding(value)

    def ciphertext_bytes(self) -> bytes:
        return b64url_decode_no_padding(self.ciphertext)

    def secret_nonce_bytes(self) -> bytes:
        return b64url_decode_no_padding(self.secret_nonce)

    def dek_nonce_bytes(self) -> bytes:
        return b64url_decode_no_padding(self.dek_nonce)

    def wrapped_dek_bytes(self) -> bytes:
        return b64url_decode_no_padding(self.wrapped_dek)

File: hbse/src/records.rs
use serde::{Deserialize, Serialize};

use crate::serialization::b64url_decode_no_padding;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SecretStatus {
    Active,
    Staged,
    Disabled,
    Destroyed,
}

impl Default for SecretStatus {
    fn default() -> Self {
        Self::Active
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SecretType {
    ApiKey,
    AccessToken,
    RefreshToken,
    Password,
    Passphrase,
    Token,
    MnemonicPhrase,
    SshKey,
    PrivateKey,
    Certificate,
    Credential,
    JsonCredential,
    Generic,
}

impl Default for SecretType {
    fn default() -> Self {
        Self::Generic
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SecretRecord {
    #[serde(default = "schema_version")]
    pub schema_version: String,
    pub vault_id: String,
    pub namespace_id: String,
    pub secret_id: String,
    pub secret_ref: String,
    pub secret_version: u64,
    #[serde(default)]
    pub status: SecretStatus,
    #[serde(default)]
    pub secret_type: SecretType,
    #[serde(default = "algorithm_id")]
    pub algorithm_id: String,
    #[serde(default = "algorithm_id")]
    pub wrap_algorithm_id: String,
    #[serde(default = "default_policy_id")]
    pub policy_id: String,
    pub policy_hash: String,
    pub metadata_hash: String,
    #[serde(default = "default_provider_policy_hash")]
    pub provider_policy_hash: String,
    pub created_at: String,
    pub secret_nonce: String,
    pub ciphertext: String,
    pub dek_nonce: String,
    pub wrapped_dek: String,
    pub secret_aad_hash: String,
    pub dek_wrap_aad_hash: String,
}

impl SecretRecord {
    pub fn ciphertext_bytes(&self) -> Result<Vec<u8>, base64::DecodeError> {
        b64url_decode_no_padding(&self.ciphertext)
    }

    pub fn secret_nonce_bytes(&self) -> Result<Vec<u8>, base64::DecodeError> {
        b64url_decode_no_padding(&self.secret_nonce)
    }

    pub fn dek_nonce_bytes(&self) -> Result<Vec<u8>, base64::DecodeError> {
        b64url_decode_no_padding(&self.dek_nonce)
    }

    pub fn wrapped_dek_bytes(&self) -> Result<Vec<u8>, base64::DecodeError> {
        b64url_decode_no_padding(&self.wrapped_dek)
    }
}

fn schema_version() -> String {
    "1".to_string()
}

fn algorithm_id() -> String {
    "AES-256-GCM".to_string()
}

fn default_policy_id() -> String {
    "default-deny".to_string()
}

fn default_provider_policy_hash() -> String {
    "unbound-provider-policy".to_string()
}
